### PR TITLE
refactor: improve usability of proto conversion traits

### PR DESCRIPTION
**Pull request by tustvold**
These extension traits grew somewhat organically as I implemented the gRPC API, and consequently are a little bit confused. This PR renames some things to make them clearer, and adds a load of documentation about how and why to use them, and also a bit about error handling in gRPC in general.

### CODE REVIEW

File: generated_types/src/google.rs [hunk 1]
```diff
@@ -310,10 +359,15 @@ impl From<PreconditionViolation> for tonic::Status {
...
+///
+/// This is useful where a field has fallible `TryFrom` conversion logic, which doesn't
+/// return [`FieldViolation`]. This will format the returned error and add the field
+/// name as context
+///
 pub trait FieldViolationExt {
     type Output;
 
-    fn field(self, field: &'static str) -> Result<Self::Output, FieldViolation>;
+    fn scope(self, field: &'static str) -> Result<Self::Output, FieldViolation>;
```
Comment (tustvold) [inline generated_types/src/google.rs:370]: This makes it consistent with the name of the member function on `FieldViolation` to further scope an error

File: generated_types/src/google.rs [hunk 2]
```diff
@@ -345,106 +399,125 @@ impl From<QuotaFailure> for tonic::Status {
...
+///
+/// * Within conversion code for a oneof enumeration
+/// * Converting from a scalar field to a custom Rust type
+///
+/// In a lot of cases, the type will be `Option<proto::Type>` or `Vec<proto::Type>`
+/// in which case `FromOptionalField` or `FromRepeatedField` should be used instead
+///
 pub(crate) trait FromField<T> {
-    fn scope(self, field: impl Into<String>) -> Result<T, FieldViolation>;
+    fn from_field(self, field: impl Into<String>) -> Result<T, FieldViolation>;
```
Comment (tustvold) [inline generated_types/src/google.rs:414]: I felt the `from_` prefix helped to identify that a type conversion is taking place

File: generated_types/src/google.rs [hunk 3]
```diff
@@ -345,106 +399,127 @@ impl From<QuotaFailure> for tonic::Status {
...
         res.map_err(|e| e.scope(field))
     }
 }
+
+/// An extension trait that adds the method `non_empty` to any `String`
+///
+/// This is useful where code wishes to require a non-empty string is specified
+///
+/// TODO: Replace with NonEmptyString type implementing TryFrom?
+pub(crate) trait NonEmptyString {
```
Comment (tustvold) [inline generated_types/src/google.rs:495]: This is FromFieldString renamed and moved lower in the file

File: generated_types/src/google.rs (hunk 2 above)
Comment (tustvold) [reply generated_types/src/google.rs:414]: Unfortunately it would appear clippy does not like it...

File: generated_types/src/google.rs [hunk 4]
```diff
@@ -345,106 +399,127 @@ impl From<QuotaFailure> for tonic::Status {
...
+///
+/// This is primarily used to define other extension traits but may be useful for:
+///
+/// * Conversion code for a oneof enumeration
+/// * Converting from a scalar field to a custom Rust type
+///
+/// In a lot of cases, the type will be `Option<proto::Type>` or `Vec<proto::Type>`
+/// in which case `FromOptionalField` or `FromRepeatedField` should be used instead
+///
+pub trait FromField<T> {
```
Comment (tustvold) [inline generated_types/src/google.rs:413]: These traits are intentionally now made public, they were only crate local whilst being iterated on

File: generated_types/src/google.rs [hunk 5]
```diff
@@ -1,5 +1,52 @@
...
+//! to abstract away concerns such as API evolution, default values, etc...
+//!
+//! Unfortunately, writing this translation code is somewhat cumbersome and so this module
+//! contains a collection of extension traits to improve the ergonomics of writing the potentially
+//! fallible conversion from the protobuf representation to the corresponding `data_types` type
+//!
+//! Each type should implement the following:
+//!
+//! * From<data_types::Type> for proto::Type
+//! * TryFrom<proto::Type, Error=FieldViolation> for data_types::Type
```
Comment (crepererum) [inline generated_types/src/google.rs:38]:
```suggestion
//! * `From<data_types::Type> for proto::Type`
//! * `TryFrom<proto::Type, Error=FieldViolation> for data_types::Type`
```

File: generated_types/src/google.rs [hunk 6]
```diff
@@ -345,106 +399,127 @@ impl From<QuotaFailure> for tonic::Status {
...
 
         res.map_err(|e| e.scope(field))
     }
 }
+
+/// An extension trait that adds the method `non_empty` to any `String`
+///
+/// This is useful where code wishes to require a non-empty string is specified
+///
+/// TODO: Replace with NonEmptyString type implementing TryFrom?
```
Comment (crepererum) [inline generated_types/src/google.rs:494]: I think that's desirable, but can wait for another PR.

**Review by crepererum — APPROVED** [review]
Thanks, I personally also struggled a bit with the extension traits in the past. The new names are definitely better.

**Review by alamb — COMMENTED** [review]
looks like a nice change to me

**Pull request merged**